### PR TITLE
diag log paging and retention

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

File: src/Microsoft.VisualStudio.Services.Agent/HostContext.cs [hunk 1]
```diff
@@ -23,33 +23,50 @@ public interface IHostContext : IDisposable
...
-                logFile,
-                FileMode.Create,
-                FileAccess.ReadWrite,
-                FileShare.Read,
-                bufferSize: 4096);
-            _traceManager = new TraceManager(new HostTraceListener(logStream), GetService<ISecretMasker>());
+            if (string.IsNullOrEmpty(logFile))
+            {
+                int logPageSize = _defaultLogPageSize;
+                string logSizeEnv = Environment.GetEnvironmentVariable($"{hostType}.logsize");
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/HostContext.cs]: lower

File: src/Microsoft.VisualStudio.Services.Agent/HostContext.cs [hunk 2]
```diff
@@ -23,33 +23,50 @@ public interface IHostContext : IDisposable
...
+                if (!string.IsNullOrEmpty(logSizeEnv))
+                {
+                    int.TryParse(logSizeEnv, out logPageSize);
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/HostContext.cs]: if ! set to default

File: src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs [hunk 3]
```diff
@@ -9,14 +9,59 @@
...
+            if (pageSizeLimit > 0)
+            {
+                _enablePageLog = true;
+                _pageSizeLimit = pageSizeLimit * 1024 * 1024;
+                _currentPageSize = 0;
+            }
+
+            if (retentionDays > 0)
+            {
+                _enableLogRollover = true;
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs]: "retention"

File: src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs [hunk 4]
```diff
@@ -89,5 +187,37 @@ private void WriteFooter(TraceEventCache eventCache)
...
+        {
+            string diagDirectory = IOUtil.GetDiagPath();
+            Directory.CreateDirectory(diagDirectory);
+            if (_enableLogRetention)
+            {
+                DirectoryInfo diags = new DirectoryInfo(diagDirectory);
+                var logs = diags.GetFiles($"{_logFilePrefix}*.log");
+                foreach (var log in logs)
+                {
+                    if (log.LastWriteTimeUtc + _retentionDays < DateTime.UtcNow)
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs]: consider .AddDays for clarity.

File: src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs [hunk 5]
```diff
@@ -33,6 +78,59 @@ public override void TraceEvent(TraceEventCache eventCache, string source, Trace
...
+            }
+        }
+
+        public override void Write(string message)
+        {
+            base.Write(message);
+            if (_enablePageLog)
+            {
+                int messageSize = UTF8Encoding.UTF8.GetByteCount(message);
+                _currentPageSize += messageSize;
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs]: This this called from multiple threads? If so, interlocked.increment

Comment (TingluoHuang) [inline src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs]: it only called from one thread, tracesource will control it.

File: src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs [hunk 6]
```diff
@@ -33,6 +78,59 @@ public override void TraceEvent(TraceEventCache eventCache, string source, Trace
...
+                Flush();
+            }
+        }
+
+        public override void Write(string message)
+        {
+            base.Write(message);
+            if (_enablePageLog)
+            {
+                int messageSize = UTF8Encoding.UTF8.GetByteCount(message);
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs]: A rough approximation may suffice as well (message.Length).

Comment (TingluoHuang) [inline src/Microsoft.VisualStudio.Services.Agent/HostTraceListener.cs]: i made this consistency with build log paging. the executioncontext log.

File: src/Microsoft.VisualStudio.Services.Agent/HostContext.cs [hunk 7]
```diff
@@ -23,33 +23,50 @@ public interface IHostContext : IDisposable
...
+            {
+                int logPageSize;
+                string logSizeEnv = Environment.GetEnvironmentVariable($"{hostType.ToLower()}.logsize");
+                if (!string.IsNullOrEmpty(logSizeEnv) || !int.TryParse(logSizeEnv, out logPageSize))
+                {
+                    logPageSize = _defaultLogPageSize;
+                }
+
+                int logRetentionDays;
+                string logRetentionDaysEnv = Environment.GetEnvironmentVariable($"{hostType.ToLower()}.logretention");
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/HostContext.cs]: tolowerinvariant would be better, although i dont think it currently matters given the actual values that are passed in.

Comment (TingluoHuang) [inline src/Microsoft.VisualStudio.Services.Agent/HostContext.cs]: base on the fact that we control the hosttype, tolower() should be fine. :D

**Comment by ericsciple** [comment]
Looks good.

**Pull request merged**